use circular buffer for connection pool

tanner0101 opened the pull request:
Changes `ConnectionPool` storage from `Array` to `CircularBuffer`.

tanner0101 commented:
Not seeing the expected results. `CircularBuffer` is ~40% slower. 

### Array 

![screen shot 2019-02-14 at 12 44 42 pm](https://user-images.githubusercontent.com/1342803/52806461-18ed3e80-3057-11e9-857e-38999876a616.png)

### Circular Buffer

![screen shot 2019-02-14 at 12 50 07 pm](https://user-images.githubusercontent.com/1342803/52806462-18ed3e80-3057-11e9-9bf1-20c13de982e6.png)

tanner0101 commented:
Story on Linux is similar, ~40% slowdown:

```
testPerformance() 
expected: 0.115 seconds
measured: 0.185
```

MrMage commented:
I guess popping/pushing the last element of the `available` array largely avoids memory allocations, anyway, so that should be fast enough. Or is this expected to be a bottleneck?

tanner0101 commented:
@MrMage the problem with the current method is the pool operates as a stack instead of a queue. Ideally we'd have FIFO behavior so that requests for a connection can't be starved by heavy usage.

MrMage commented:
@tanner0101 but how would the underlying buffer implementation change that behavior? If you mean with FIFO behavior that "the first to request a connection should receive the next available connection", I don't think that behavior relates to the actual implementation of the buffer structure. 

The choice of a ring buffer only ensures (I think) that all connections get used with roughly the same frequency (instead of mostly using just one connection), which shouldn't make a difference in practice.

tanner0101 commented:
The problem is that using array as a queue will result in `O(n)` performance. You need to either:
- prepend O(n)  + popLast O(1)
- append O(1) + popFirst O(n)

With circular buffer, you get:
- prepend O(1) + removeLast O(1) (amortized) 
- append O(1) + removeFirst O(1) (amortized)

MrMage commented:
But what's the problem with doing append+popLast on the array? Both have O(1) complexity.

> On 14. Feb 2019, at 20:55, Tanner <notifications@github.com> wrote:
> 
> The problem is that using array as a queue will result in O(n) performance. You need to either:
> 
> prepend O(n) + popLast O(1)
> append O(1) + popFirst O(n)
> With circular buffer, you get:
> 
> prepend O(1) + removeLast O(1) (amortized)
> append O(1) + removeFirst O(1) (amortized)
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

calebkleveter commented:
@MrMage Because that results in FILO, which while it works, will use some connections a lot while leaving other ones untouched. We just want to balance the load across all the connections.

MrMage commented:
Thank you for the elaboration. But what is the tangible benefit of balancing the load across all connections?

> On 15. Feb 2019, at 13:09, Caleb Kleveter <notifications@github.com> wrote:
> 
> @MrMage Because that results in FILO, which while it works, will use some connections a lot while leaving other ones untouched. We just want to balance the load across all the connections.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

tanner0101 commented:
@MrMage @calebkleveter balancing the connections is not the problem actually. It's the _requests_ for connections that are being stored here, and that should be FIFO. Imagine this case:

- User A makes a request for a connection, gets one connection 1.
- User B makes a request for a connection, they are exhausted so he gets a promise.
- User A makes another request for a connection, they are still exhausted so he gets a promise.
- User A returns connection 1. His previous request is fulfilled with connection 1. 
- User B is still waiting for a connection.

MrMage commented:
@tanner0101 that makes more sense, but as I explained in https://github.com/vapor/nio-kit/pull/25#issuecomment-463734852 that problem is independent of the choice of buffer structure; instead you need a queue (or ring buffer) structure for the *waiting requests*, not the *available connections*.

tanner0101 commented:
![image](https://user-images.githubusercontent.com/1342803/52889106-7fa05400-314c-11e9-9069-6332e474423d.png)

Thank you for catching that @MrMage.

I've updated the PR now with FIFO for _waiters_ as it should have been. Performance actually improves slightly which is great since this is technically a bug fix. 

![screen shot 2019-02-15 at 6 07 47 pm](https://user-images.githubusercontent.com/1342803/52889140-9d6db900-314c-11e9-90a1-dd83dbc22bfa.png)

I've also added a test to ensure waiters are fulfilled FIFO.

tanner0101 commented:
Results on Linux are even better:

```
[PERFORMANCE] testPerformance() expected: 0.115 seconds
Test Case 'ConnectionPoolTests.testPerformance' measured average: 0.088
```

MrLotU commented on Sources/NIOKit/ConnectionPool.swift:
```diff
@@ -83,8 +83,9 @@ public final class ConnectionPool<Source> where Source: ConnectionPoolSource  {
         self.config = config
         self.source = source
         self.available = []
+        self.available.reserveCapacity(config.maxConnections)
```
Out of curiosity, why did we add this? :)

tanner0101 replied:
The array should never need to hold more than that capacity, so reserving should prevent any need for array re-alloc during usage.

MrLotU replied:
Thought that was it, thanks!

penny-coin commented:
Hey @tanner0101, you just merged a pull request, have a coin! 

You now have 1093 coins.
